hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -275,4 +307,44 @@ protected Pair<Map<ConfigProperty, String>, List<ConfigProperty>> downgrade(Hood
       throw new HoodieUpgradeDowngradeException(fromVersion.versionCode(), toVersion.versionCode(), false);
     }
   }
+
+  /**
+   * Checks if any handlers in the upgrade/downgrade path need compaction and performs it once before starting.
+   * This ensures compaction happens only when needed at the very beginning of the process.
+   *
+   * @param fromVersion the current table version
+   * @param toVersion   the target table version
+   */
+  private boolean requiresCompaction(HoodieTableVersion fromVersion, HoodieTableVersion toVersion) {
```

Comment:
why can't we pass the `isUpgrade` boolean from above as an argument here and avoid L325

Reply:
Yea let me do that as well.